### PR TITLE
GH-1125: Add DeclarableCustomizer

### PR DESCRIPTION
**Pull request by garyrussell**
Resolves https://github.com/spring-projects/spring-amqp/issues/1125

- pull up the common `arguments` field to `AbstractDeclarable`
- add `x-single-active-consumer` as a first class `QueueBuilder` argumemt

### CODE REVIEW

File: spring-amqp/src/main/java/org/springframework/amqp/core/Declarable.java [hunk 1]
```diff
@@ -50,4 +52,24 @@
...
+	}
+
+	/**
+	 * Remove an argument from the declarable.
+	 * @param name the argument name.
+	 * @return the argument value or null if not present.
+	 * @since 2.2.2
+	 */
+	default @Nullable Object removeArgument(String name) {
+		return false;
```
Comment (artembilan) [inline spring-amqp/src/main/java/org/springframework/amqp/core/Declarable.java]: `false`? Why not `null`?

File: spring-amqp/src/main/java/org/springframework/amqp/core/DeclarableCustomizer.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+/**
+ * Beans of this type are invoked by the {@link AmqpAdmin} before declaring the
+ * {@link Declarable}, allowing customization thereof.
+ *
+ * @author Gary Russell
+ * @since 2.2.2
+ *
+ */
+@FunctionalInterface
+public interface DeclarableCustomizer extends Function<Declarable, Declarable> {
```
Comment (artembilan) [inline spring-amqp/src/main/java/org/springframework/amqp/core/DeclarableCustomizer.java]: Why `function`, but not `consumer`?
Are you intended to replace property with result of this customizer?

File: spring-amqp/src/main/java/org/springframework/amqp/core/AbstractDeclarable.java [hunk 3]
```diff
@@ -38,6 +41,22 @@
...
 	private boolean ignoreDeclarationExceptions;
 
+	private final Map<String, Object> arguments;
+
+	/**
+	 * Construct an instance with the supplied arguments, or an empty map if null.
+	 * @param arguments the arguments.
+	 * @since 2.2.2
+	 */
+	public AbstractDeclarable(@Nullable Map<String, Object> arguments) {
```
[the remark this answers is not in the document]
Comment (garyrussell) [reply spring-amqp/src/main/java/org/springframework/amqp/core/AbstractDeclarable.java]: I seriously doubt there are custom subclasses but, yes, I'll add back the default ctor.

File: spring-amqp/src/main/java/org/springframework/amqp/core/Declarable.java (hunk 1 above)
Comment (garyrussell) [reply spring-amqp/src/main/java/org/springframework/amqp/core/Declarable.java]: Oops

File: spring-amqp/src/main/java/org/springframework/amqp/core/DeclarableCustomizer.java (hunk 2 above)
Comment (garyrussell) [reply spring-amqp/src/main/java/org/springframework/amqp/core/DeclarableCustomizer.java]: Yes; they can replace the object if they need to for some reason (aside from the arguments, they are mostly immutable).

**Pull request merged**